Directory structure:

-During the sequencing run, the MinKnow software uses 'fast-basecalling' to convert fast5 raw data -> fastqs. In doing this, it also applies some QC to determine if reads 'pass' or 'fail.' So during the run, fast-basecalled raw fast5 files are classified as pass or fail and saved in the fast5_pass and fast5_fail directories. Generated fastqs are saved in the fastq_pass and fastq_fail directories. 
-Even though it is fast, fast basecalling still takes longer than sequencing, so when the run finishes, MinKnow doesn't get to all the fast5 files. these are saved in fast5_skip.
-After sequencing, I went back and re-basecalled ALL of the raw fast5s (pass, fail, and skip) with both high-accuracy and super-accuracy basecalling, with calibration strand detection enabled.
  -fastq_hac_cd: fastq's generated by high accuracy basecalling.
    1. pass
        a.fastqs that pass based on minimum qscore = 10
        b.split_pass: contains fastqs with reads that have been split. Fastqs also contain a copy of unsplit reads 
    2. fail: fastqs that fail based on minimum qscore = 10
    3. calibration_strands: fastqs with identified calibration strands
  
  -fastq_sac_cd: fastq's generated by super accuracy basecalling.
    1. pass
        a.fastqs that pass based on minimum qscore = 10
        b.split_pass: contains fastqs with reads that have been split. Fastqs also contain a copy of unsplit reads 
    2. fail: fastqs that fail based on minimum qscore = 10
    3. calibration_strands: fastqs with identified calibration strands


high-accuracy with calibration strand detection:
```{bash}
#skipped fast5s
~/ont-guppy/bin/guppy_basecaller --input_path /mnt/data0/MinION_reads/JepsonT/JepsonT/20220126_1311_MN23913_FAR20182_271593e5/fast5_skip/ --save_path /mnt/data0/MinION_reads/JepsonT/JepsonT/20220126_1311_MN23913_FAR20182_271593e5/fastq_hac_calib --calib_detect --config dna_r10.4_e8.1_hac.cfg -x cuda:0 --num_callers 8

#passed fast5s
~/ont-guppy/bin/guppy_basecaller --input_path /mnt/data0/MinION_reads/JepsonT/JepsonT/20220126_1311_MN23913_FAR20182_271593e5/fast5_pass/ --save_path /mnt/data0/MinION_reads/JepsonT/JepsonT/20220126_1311_MN23913_FAR20182_271593e5/fastq_hac_calib --calib_detect --config dna_r10.4_e8.1_hac.cfg -x cuda:0 --num_callers 8

#failed fast5s
~/ont-guppy/bin/guppy_basecaller --input_path /mnt/data0/MinION_reads/JepsonT/JepsonT/20220126_1311_MN23913_FAR20182_271593e5/fast5_fail/ --save_path /mnt/data0/MinION_reads/JepsonT/JepsonT/20220126_1311_MN23913_FAR20182_271593e5/fastq_hac_calib --calib_detect --config dna_r10.4_e8.1_hac.cfg -x cuda:0 --num_callers 8
```

Software used: 

-ONT Guppy: basecalling. (protocol available on ONT website)
-ONT duplex_tools split_on_adapter (https://github.com/nanoporetech/duplex-tools)


Code:

super accuracy basecalling with calibration strand detection:

```{bash}
~/ont-guppy/bin/guppy_basecaller --input_path /mnt/data0/MinION_reads/JepsonT/JepsonT/20220126_1311_MN23913_FAR20182_271593e5/fast5_skip/ --save_path /mnt/data0/MinION_reads/JepsonT/JepsonT/20220126_1311_MN23913_FAR20182_271593e5/fastq_sac_calib --calib_detect --config dna_r10.4_e8.1_sup.cfg -x cuda:0 --num_callers 8
```
output:
(kraken) user@user-System-Product-Name:/mnt/data0/MinION_reads/JepsonT/JepsonT/20220126_1311_MN23913_FAR20182_271593e5$ ~/ont-guppy/bin/guppy_basecaller --input_path /mnt/data0/MinION_reads/JepsonT/JepsonT/20220126_1311_MN23913_FAR20182_271593e5/fast5_skip/ --save_path /mnt/data0/MinION_reads/JepsonT/JepsonT/20220126_1311_MN23913_FAR20182_271593e5/fastq_sac_calib --calib_detect --config dna_r10.4_e8.1_sup.cfg -x cuda:0 --num_callers 8
ONT Guppy basecalling software version 5.0.11+2b6dbff
config file:        /home/user/ont-guppy/data/dna_r10.4_e8.1_sup.cfg
model file:         /home/user/ont-guppy/data/template_r10.4_e8.1_sup.jsn
input path:         /mnt/data0/MinION_reads/JepsonT/JepsonT/20220126_1311_MN23913_FAR20182_271593e5/fast5_skip/
save path:          /mnt/data0/MinION_reads/JepsonT/JepsonT/20220126_1311_MN23913_FAR20182_271593e5/fastq_sac_calib
chunk size:         2000
chunks per runner:  256
minimum qscore:     10
records per file:   4000
num basecallers:    8
gpu device:         cuda:0
kernel path:
runners per device: 12
Found 3348 fast5 files to process.
Init time: 1888 ms

0%   10   20   30   40   50   60   70   80   90   100%
|----|----|----|----|----|----|----|----|----|----|
***************************************************
Caller time: 47951328 ms, Samples called: 189199682264, samples/s: 3.94566e+06
Finishing up any open output files.
Basecalling completed successfully.

```{bash}
#passed fast5s
 ~/ont-guppy/bin/guppy_basecaller --input_path /mnt/data0/MinION_reads/JepsonT/JepsonT/20220126_1311_MN23913_FAR20182_271593e5/fast5_pass/ --save_path /mnt/data0/MinION_reads/JepsonT/JepsonT/20220126_1311_MN23913_FAR20182_271593e5/fastq_sac_calib --calib_detect --config dna_r10.4_e8.1_sup.cfg -x cuda:0 --num_callers 8
```
output:
ONT Guppy basecalling software version 5.0.11+2b6dbff
config file:        /home/user/ont-guppy/data/dna_r10.4_e8.1_sup.cfg
model file:         /home/user/ont-guppy/data/template_r10.4_e8.1_sup.jsn
input path:         /mnt/data0/MinION_reads/JepsonT/JepsonT/20220126_1311_MN23913_FAR20182_271593e5/fast5_pass/
save path:          /mnt/data0/MinION_reads/JepsonT/JepsonT/20220126_1311_MN23913_FAR20182_271593e5/fastq_sac_calib
chunk size:         2000
chunks per runner:  256
minimum qscore:     10
records per file:   4000
num basecallers:    8
gpu device:         cuda:0
kernel path:
runners per device: 12
Found 526 fast5 files to process.
Init time: 1902 ms

0%   10   20   30   40   50   60   70   80   90   100%
|----|----|----|----|----|----|----|----|----|----|
***************************************************
Caller time: 6564657 ms, Samples called: 25936161286, samples/s: 3.95088e+06
Finishing up any open output files.
Basecalling completed successfully.


```{bash}
#failed fast5s
~/ont-guppy/bin/guppy_basecaller --input_path /mnt/data0/MinION_reads/JepsonT/JepsonT/20220126_1311_MN23913_FAR20182_271593e5/fast5_fail/ --save_path /mnt/data0/MinION_reads/JepsonT/JepsonT/20220126_1311_MN23913_FAR20182_271593e5/fastq_sac_calib --calib_detect --config dna_r10.4_e8.1_sup.cfg -x cuda:0 --num_callers 8
```
output:
ONT Guppy basecalling software version 5.0.11+2b6dbff
config file:        /home/user/ont-guppy/data/dna_r10.4_e8.1_sup.cfg
model file:         /home/user/ont-guppy/data/template_r10.4_e8.1_sup.jsn
input path:         /mnt/data0/MinION_reads/JepsonT/JepsonT/20220126_1311_MN23913_FAR20182_271593e5/fast5_fail/
save path:          /mnt/data0/MinION_reads/JepsonT/JepsonT/20220126_1311_MN23913_FAR20182_271593e5/fastq_sac_calib
chunk size:         2000
chunks per runner:  256
minimum qscore:     10
records per file:   4000
num basecallers:    8
gpu device:         cuda:0
kernel path:
runners per device: 12
Found 214 fast5 files to process.
Init time: 1839 ms

0%   10   20   30   40   50   60   70   80   90   100%
|----|----|----|----|----|----|----|----|----|----|
***************************************************
Caller time: 2188507 ms, Samples called: 8681321524, samples/s: 3.96678e+06
Finishing up any open output files.
Basecalling completed successfully.



Splitting reads with ONT duplex_tools split_on_adapter (https://github.com/nanoporetech/duplex-tools)

```{bash}
#hac fastqs
duplex_tools split_on_adapter /mnt/data0/MinION_reads/JepsonT/JepsonT/20220126_1311_MN23913_FAR20182_271593e5/fastq_hac_calib/pass /mnt/data0/MinION_reads/JepsonT/JepsonT/20220126_1311_MN23913_FAR20182_271593e5/fastq_hac_calib/pass/split_pass Native
```
#output: 
837it [00:33, 24.85it/s]
Split 342340 reads kept 2185710 reads

```{bash}
#sac fastqs
duplex_tools split_on_adapter /mnt/data0/MinION_reads/JepsonT/JepsonT/20220126_1311_MN23913_FAR20182_271593e5/fastq_sac_calib/pass /mnt/data0/MinION_reads/JepsonT/JepsonT/20220126_1311_MN23913_FAR20182_271593e5/fastq_sac_calib/pass/split_pass Native
```
#output: 
837it [00:35, 23.40it/s]
Split 366495 reads kept 2197604 reads



Calibration strand analysis:

Info on the calibration strand detection and analysis is in the file sequencing_summary inside the parent fastq_hac_cd and fastq_sac_cd folders
sequencing_summary.txt has a line with info for every read. 
It appears that if the calibration strand was detetected within a read, metrics related to the calibration strand are calculated and included in the later columns of this file
it appears that if the calibration strand is NOT detected within a read, the value for all calibration-related columns is '-1.' However, I cannot find documentation for this.
you can sort these files by 'calibration_strand_identity' or 'calibration_strand_accuracy'and calculate averages for reads assigned a non -1 value

I went ahead and did this for JepsonT but you can repeat to see if you get the same numbers if you would like (numbers below are all averages of non -1 values for the HAC and SAC basecalled reads)
filename	total_reads	calibration_strand_genome	calibration_strand_num_insertions	calibration_strand_num_deletions	calibration_strand_num_aligned	calibration_strand_num_correct	calibration_strand_identity	calibration_strand_accuracy	calibration_strand_score
hac	593	Lambda_3.6kb	120.2124789	241.5649241	3127.268128	2820.325464	0.899507486	0.806583899	0
sac	889	Lambda_3.6kb	132.6085489	231.5714286	3136.388076	2812.475816	0.894538325	0.802577209	0

for each read:
accuracy = # bases correct/(# insertions + # deletions + # bases aligned)
identity = # bases correct/ # bases aligned 

So looks like super accuracy basecalling identified more calibration strands, but resulted in pretty similar average accuracy and identity.



